src/lib.rs

```diff
@@ -40,188 +46,125 @@
-impl<State: Clone + Send + Sync + 'static> Middleware<State> for OpenTelemetryTracingMiddleware {
+impl<T: Tracer + Send + Sync, State: Clone + Send + Sync + 'static> Middleware<State>
+    for OpenTelemetryTracingMiddleware<T>
+{
     async fn handle(&self, req: Request<State>, next: Next<'_, State>) -> Result {
-        let mut req_headers = std::collections::HashMap::new();
-        for (k, v) in req.iter() {
-            req_headers.insert(k.to_string(), v.last().to_string());
+        let method = req.method().clone();
+        let url = req.url().clone();
+
+        let mut attributes = vec![
+            trace::HTTP_METHOD.string(method.to_string()),
+            trace::HTTP_SCHEME.string(url.scheme().to_owned()),
+            trace::HTTP_URL.string(url.to_string()),
+            trace::HTTP_TARGET.string(http_target(&url)),
+        ];
+
+        attributes.reserve(6);
+
+        if let Some(version) = req.version() {
+            attributes.push(trace::HTTP_FLAVOR.string(http_version_str(version)));
         }
 
-        let b3_propagator = B3Propagator::with_encoding(B3Encoding::SingleAndMultiHeader);
-        let trace_context_propagator = TraceContextPropagator::new();
-        let composite_propagator =
-            HttpTextCompositePropagator::new(vec![Box::new(b3_propagator), Box::new(trace_context_propagator)]);
-        let _parent = composite_propagator
-            .extract_with_context(&Context::current(), &req_headers)
-            .attach();
-        let tracer = global::tracer("opentelemetry-tide");
-        let span_name = format!("{} {}", req.method(), req.url());
-        let mut builder = tracer.span_builder(&span_name);
-        builder.span_kind = Some(SpanKind::Server);
+        if let Some(host) = url.host_str() {
+            attributes.push(trace::HTTP_HOST.string(host.to_owned()));
+        }
 
-        let url = req.url();
-        let attributes = vec![
-            KeyValue::new(HTTP_METHOD_ATTRIBUTE, req.method().to_string()),
-            KeyValue::new(HTTP_FLAVOR_ATTRIBUTE, http_version_str(req.version())),
-            KeyValue::new(HTTP_SCHEME_ATTRIBUTE, url.scheme()),
-            KeyValue::new(HTTP_URL_ATTRIBUTE, url.as_str()),
-            KeyValue::new(HTTP_HOST_ATTRIBUTE, url.host_str().unwrap_or(UNKNOWN)),
-            KeyValue::new(HTTP_TARGET_ATTRIBUTE, http_target(url)),
-            KeyValue::new(HTTP_SERVER_NAME_ATTRIBUTE, url.domain().unwrap_or(UNKNOWN)),
-            KeyValue::new(
-                NET_HOST_PORT_ATTRIBUTE,
-                Value::U64(url.port_or_known_default().unwrap_or(0u16) as u64),
-            ),
-            KeyValue::new(HTTP_CLIENT_IP_ATTRIBUTE, net_addr_ip(req.peer_addr())),
-            KeyValue::new(NET_PEER_IP_ATTRIBUTE, net_addr_ip(req.remote())),
-        ];
-        builder.attributes = Some(attributes);
-        let span = tracer.build(builder);
-        let _guard = tracer.mark_span_as_active(span);
+        if let Some(domain) = url.domain() {
```

Comment:
I decided to do conditionals so that tags only exist if they have a useful value.

Reply:
A little bit more noise on the rust side but far less noise in the logs